apps/core/faviconmanager.cpp

```diff
@@ -0,0 +1,219 @@
+/****************************************************************************
+**
+** Copyright (c) 2021 Open Mobile Platform LLC.
+**
+****************************************************************************/
+
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this file,
+ * You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+#include <QDebug>
+#include <QFile>
+#include <QTextStream>
+#include <QJsonObject>
+#include <QJsonArray>
+#include <QJsonDocument>
+#include <QUrl>
+#include <memory>
+
+#include "browserpaths.h"
+#include "datafetcher.h"
+
+#include "faviconmanager.h"
+
+FaviconManager::FaviconManager(QObject* parent)
+    : QObject(parent)
+    , m_faviconSets()
+{
+}
+
+FaviconManager* FaviconManager::instance()
+{
+    static FaviconManager* singleton;
+    if (!singleton) {
+        singleton = new FaviconManager();
+    }
+
+    return singleton;
+}
+
+static QString sanitiseHostname(const QString &hostname)
+{
+    QUrl url(hostname);
+    // Maybe port should be included too
+    QString sanitised = QString("%1://%2").arg(url.scheme()).arg(url.host());
```

Comment:
Maybe it's better just to return this object here without any extra variable introduction.

Reply:
Unfortunately some form of sanitization is needed, because the logins are stored with full URLs, so hostnames are needed for lookup. It could be moved oustide of FaviconManger though?

Reply:
I mean just this:
```suggestion
    return QString("%1://%2").arg(url.scheme()).arg(url.host());
```

Reply:
Ah, I'm so sorry, I completely misunderstood. Yes, that's a good idea: I've made this change.